sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -60,21 +60,19 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
     session.sharedState.cacheManager.recacheByPlan(session, r)
   }
 
+  private def recacheTable(r: ResolvedTable)(): Unit = {
+    val v2Relation = DataSourceV2Relation.create(r.table, Some(r.catalog), Some(r.identifier))
+    session.sharedState.cacheManager.recacheByPlan(session, v2Relation)
+  }
```

Comment:
This seems the main point of this change. And as other comments, this looks like a behavior change or an inconsistent behavior to v1.

> Reduce the number of calls to the Cache Manager when need to recache a table. Before the changes, invalidateCache() invokes the Cache Manager 3 times: lookupCachedData(), uncacheQuery() and cacheQuery().

Hmm, does this stand for the reason of this change? If we don't consider the behavior change part, `recacheByPlan` is actually more complicated than other three calls `lookupCachedData`, `uncacheQuery` and `cacheQuery`.

Reply:
I also think should be considered as a behavior change. Even though query result remains the same, it now lazily caches data which changes query execution time and memory/storage consumption. Because this, we should make a similar change for v1 as well.

And please update the PR description to make it more clear (e.g., user-facing change).

Reply:
> this looks like a behavior change ...

From correctness point of view, there are no behavior change.

> an inconsistent behavior to v1

Let's look at the command affected in this PR:
- v1 `ALTER TABLE .. DROP PARTITION` does not do re-caching in v3.0, and has a bug, see https://github.com/apache/spark/pull/31006
- v1 `ALTER TABLE .. ADD PARTITION` does not do re-caching too, and has a correctness bug in 3.0. See the fix https://github.com/apache/spark/pull/31116
- v1 `ALTER TABLE .. RENAME PARTITION` also has correctness bug in 3.0: https://github.com/apache/spark/pull/31060

Could you, please, explain what do you mean by "inconsistent" behavior to v1? Inconsistent to recent fixes?

> does this stand for the reason of this change?

Not main reason but one of the reasons.

> recacheByPlan is actually more complicated than other three calls

This is arguable. I do believe it is simpler more efficient.

Reply:
> ... we should make a similar change for v1 as well.

I agree. We should change cache refreshing in v1 command either by introducing new function for that or modifying `CatalogImpl.refreshTable()`: https://github.com/apache/spark/blob/62d82b5b270c2eea27ba026e66b7d6598bbaa0a6/sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala#L538-L558
which can be used for v1 as well as for v2 tables/views (potentially but I am not sure that it works for v2 tables for now).

Reply:
Doesn't this only change for v2 behavior: https://github.com/apache/spark/pull/31172#discussion_r557031365? Based on https://github.com/apache/spark/pull/31172#discussion_r557030832, if we just uncache dependent caches for v1, isn't inconsistent to v1?

> This is arguable. I do believe it is simpler more efficient.

If someone creates a method in cache manager to include three calls? :) I don't think the number of calls here is the point .

BTW, I think recaching dependent caches is the most important point of this change, but the PR description doesn't explicitly mention it. This should be clear in the description.

When I'm writing the comment, I saw you update the title and it looks more precise. That's good, but please update the description too.

Reply:
> If someone creates a method in cache manager to include three calls? :) I don't think the number of calls here is the point .

Those 2 back-to-back calls:
```scala
    val cache = session.sharedState.cacheManager.lookupCachedData(v2Relation)
    session.sharedState.cacheManager.uncacheQuery(session, v2Relation, cascade = true)
```
lead to unnecessary traversals over cache. In total 3 calls performs more work in the cache, I do believe. 

> That's good, but please update the description too.

I have updated it. Please, review it.

Reply:
yea let's update v1 tables as well (with another PR). It's weird if my query gets uncached because someone else refreshes a table.